LE audio CAP acceptor and initiator API

Thalley opened the pull request:
Suggested API for the CAP acceptor and initiator roles. 

The CAP acceptor is a mostly autonomous entity once the values have been registered with the unicast server (and PACS/ASCS) API, and does not warrant a very large API. 

Furthermore, the Find CCID service procedure is also used for the commander, so that is a common procedure. 

The CAP initiator mostly builds on the BAP unicast and broadcast source APIs. 

fixes https://github.com/zephyrproject-rtos/zephyr/issues/41590
fixes https://github.com/zephyrproject-rtos/zephyr/issues/41186

Thalley committed:
Bluetooth: Audio: Initial API for CAP acceptor role

Adds the initial API for the CAP acceptor role.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented:
Can't add you as reviewers, but also ping @rymanluk and @szymon-czapracki

Thalley commented:
This PR also currently depends on files and/or structs not yet upstreamed, but either exists in PRs or in the topic-le-audio branch.

rymanluk commented on include/bluetooth/audio/cap.h:
```diff
@@ -0,0 +1,97 @@
+ *
+ * @return 0 if success, errno on failure.
+ */
+int bt_cap_find_ccid_service_service(struct bt_conn *conn, uint8_t ccid);
```
Should't it be ccid_type instead? How the user knows ccid ?

Thalley replied:
I am not sure I understand what you mean by `ccid_type`. 

The CCID value will be part of the metadata when setting up a stream as far as I can tell, and is provided the client setting up the stream. 

This is something that will need to be exposed with the unicast server API I suppose, but that part is still missing from that API. 

One thing I am also a bit unclear on, is how much of the API we should duplicate, e.g. if there's a call to a unicast server callback (pending change in the topic-le-audio branch), should the unicast server call a similar callback for the CAP acceptor? We currently do that for e.g. the ISO callbacks, which are currently duplicated by the audio stream implementation. 

I think we might want to add a metadata updated callback for the CAP acceptor, which would include the metadata itself, and thus the UUID, possibly deconstructed from the LTV list and put into a user-friendly struct instead. 

Let me add an update for that, and see if that makes sense.

Thalley replied:
I'll wait with an (possible) update until we have resolved whether or not we want to duplicate the callbacks for CAP, when they already exists for BAP.

rymanluk replied:
Sorry for the shortcut, but maybe I misunderstood API here. 

When reading this PR, I see that it is planned that CAP is exposing context for TBS and MCS to the User. User can find the context for given bt_conn and ccid. Is that correct?

Indeed User on the Acceptor side might be interested in CCID for TBS and MCS and the context for those services, in order to do operations on that and also possible to have additional knowledge about who is controlling stream (as you pointed metadata in BAP).
So when looking into this, I thought that idea is that CAP will discover those services and make it easy to user to get those.
Having that in mind, I tried to understand how the user will get CCID to provide to `bt_cap_find_ccid_service_service`, as the only way is to actually read characteristic on TBS or MCS (depends which you are interested in). 

So I though, maybe it shall be  `bt_cap_ccid_record_type` instead?

Thalley replied:
Sorry for late reply, must've missed your comments @rymanluk 

> When reading this PR, I see that it is planned that CAP is exposing context for TBS and MCS to the User. User can find the context for given bt_conn and ccid. Is that correct?

Correct, the find_by_ccid is a spec defined procedure for the acceptor to match the CCID from the BAP metadata to a MCS or TBS instance on the initiator. 

> So when looking into this, I thought that idea is that CAP will discover those services and make it easy to user to get those.

The CCID of those services are, as far as I know, not constant and may change. That being said, it is probably possible to have the acceptor (and commandar) implementation autonomously do this, which would make the acceptor 100% autonomous regarding CAP procedures (but not BAP procedures). 

> Having that in mind, I tried to understand how the user will get CCID to provide to bt_cap_find_ccid_service_service, as the only way is to actually read characteristic on TBS or MCS (depends which you are interested in).

The CCID will be in the metadata from BAP

> So I though, maybe it shall be bt_cap_ccid_record_type instead?

Not sure what you mean by that. Are you suggesting to rename the function? 

> Do we need to expose BAP to the User since we do expose CAP? I thought CAP will do all we need (using BAP)

As discussed in the meeting, we don't want to duplicate callbacks and functions (procedures) from other profile implementations, so to use CAP, the user will also need to register for the BAP callbacks.

rymanluk replied:
@Thalley we are good as agreed on the meeting, that idea is the CCID will be provided by metadata and then user will use it for the specific context.

Thalley commented:
One thing we need to consider for the CAP API, is whether CAP should encapsulated all the APIs of the lower layers (e.g. BAP, VCP, etc.) and be a single point of entry for the application/upper layers, or if CAP should only expose CAP specific values and functions, and depend on the application also using the lower layer APIs, e.g. Option A or Option B in the following figure. 

![image](https://user-images.githubusercontent.com/2520413/145824223-ce0e98b7-f8b6-468b-a752-9dda243cb20f.png)

Casper-Bonde-Bose commented:
I tend to favor option A - to ease maintenance. I also think Option A forces app-developers to think about supporting clients that do not have CAP implemented.
If Option B is selected the API parameters should be the same as used for BAP (where it makes sense) - I think all implementations will need to also support non-CAP based operation, and if the same parameters are used, it is easy to build an application that uses both API's.

With solution B I do see an issue if multiple API's expose a way to register endpoints/capabilities - if that is the case it either needs to be very well documented how to use it - or the CAP/BAP needs to merge capabilities registered through different interfaces and abstract that from the user.

Thalley commented:
> One thing we need to consider for the CAP API, is whether CAP should encapsulated all the APIs of the lower layers (e.g. BAP, VCP, etc.) and be a single point of entry for the application/upper layers, or if CAP should only expose CAP specific values and functions, and depend on the application also using the lower layer APIs, e.g. Option A or Option B in the following figure.
> 
> ![image](https://user-images.githubusercontent.com/2520413/145824223-ce0e98b7-f8b6-468b-a752-9dda243cb20f.png)

Option A seems to be the consensus.

Thalley committed:
Bluetooth: Audio: CAP initiator API suggestion

Add a suggestion to the CAP inititator API. This builds
on the audio.h API, and attempts to reuse as much as possible.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

asbjornsabo commented on include/bluetooth/audio/cap.h:
```diff
@@ -0,0 +1,352 @@
+
+/** Type of CCID records */
+enum bt_cap_ccid_record_type {
+	/** CCID belongs to a Telephone Bearer Service instance */
```
Dummy comment

rymanluk commented on include/bluetooth/audio/cap.h:
```diff
@@ -70,6 +70,57 @@ struct bt_cap_cb {
+	 *                       occurred. NULL if @p err is 0.
+	 */
+	void (*unicast_start_complete)(struct bt_audio_unicast_group *unicast_group,
+				       int err, struct bt_conn *conn);
```
is this callback called multiple times per each member in the group CAP is starting?

Thalley replied:
Hmm, the idea was to only call it once for either success (for all devices) or failure (for any device). Do you think it would be better to periodically update the application with the status?

rymanluk replied:
I'm asking because there is `struct bt_conn *conn` parameter which makes me think it is per device. Calling it once should be enough but bt_conn is not needed, as it should be a list of streams inside the group. Is that correct?

Thalley replied:
The `conn` parameter is supplied here to provide information about which device was the cause of the failure, as per the documentation:
```
	 * @param conn           Pointer to the connection where the error
	 *                       occurred. NULL if @p err is 0.
```

I figured it would be useful for the application to know which device caused the issue, e.g. if that specific device disconnected during this procedure, so that the application can perform a reconnect or remove it from the set and restart the procedure.

rymanluk commented on include/bluetooth/audio/cap.h:
```diff
@@ -86,6 +137,210 @@ struct bt_cap_cb {
+	 * stream[i] will be associated with members[i] if not already
+	 * initialized, else the stream will be verified against the member.
+	 */
+	struct bt_audio_stream **streams;
```
Could you put a light  on how this is going to be used?  In what state `stream` is provided here?

Also I see that `struct bt_audio_stream` containst `struct bt_conn` and `struct bt_audio_ep`, so maybe in the end we don't need the above union at all as list of streams actually defines the group?

Thalley replied:
> In what state stream is provided here?

Can you elaborate on what you mean by `state` here? 

> Also I see that struct bt_audio_stream containst struct bt_conn and struct bt_audio_ep, so maybe in the end we don't need the above union at all as list of streams actually defines the group?

`struct bt_audio_stream` is application allocated and does not (or should not) have any application-assigned values for `bt_conn` or `bt_audio_ep` (that is not very clear from the API as of now, I agree). This is probably originally been designed based on the `l2cap` API, which uses the same formula so to say (which possibly isn't a very good one, since it is easy for the application to modify these values and cause issues). 

So `struct bt_audio_stream` contains some applicaiton-assigned values, and then a couple of values that the stack, i.e. CAP/BAP will fill in, such as the `bt_conn` and `bt_audio_ep`.

rymanluk replied:
Need to agree this is confusing. For example, how user can say create request for following usecase
 - 1 Sink stream to Device A
 - 1 Source stream to Device A
 - 1 Sink stream to Device B
 ?
All in one CIG of course.

Thalley replied:
Hmm, that is a good point. 

To achieve that the application would need to do
```
members[0] = device_1;
streams[0] = device_1_sink_stream;
members[0] = device_1;
streams[0] = device_1_source_stream;
members[0] = device_2;
streams[0] = device_2_sink_stream;
```

and then `count` should mainly refer to the number of streams. 

Might need to rethink this to see if there's a better way to avoid multiple 

The alternative is to couple the streams and member pointers in a preliminary function, and then just provide the list of streams here. 

Please let me know if you have any suggestions :)

rymanluk replied:
I think  I would suggest that CAP has high level API here. Eg.
1) User provides to CAP list of members which are involved in the scenario. (as it is now)
2) User provides to CAP information what scenario should be realized by CAP.  E.g number of Sink connection per device and number of source connections. Or maybe even just a stream context type which would match to some policy inside the CAP e.g. 
    a) for MEDIA, CAP will establish uni direction CISes to all the devices in *members 
    b) For Phone, CAP will  establish bi-directional CISes to all the devices in *members.
3) user could provide target latency and required sample frequency to Sink and Source streams.
4) CAP would take care for allocation memory for `struct bt_audio_stream`  per device (not sure if possible) while creating streams

Thalley replied:
> 2. User provides to CAP information what scenario should be realized by CAP

I think it makes sense to use the codec config data and metadata to retrieve as much information about the scenario as possible. The details is not something I've considered too much yet. 


> 3. user could provide target latency and required sample frequency to Sink and Source streams.

Agree, that will be part of the `qos` parameter here

> 4. CAP would take care for allocation memory for struct bt_audio_stream per device (not sure if possible) while creating streams

This one I am not so sure about. I am not opposed to allocating the streams in CAP, but I also do not see the benefit. Can you elaborate on what your thoughts here are?

rymanluk commented on include/bluetooth/audio/cap.h:
```diff
@@ -86,6 +137,210 @@ struct bt_cap_cb {
+	const struct bt_codec *codec;
+
+	/** Quality of Service configuration */
+	const struct bt_codec_qos *qos;
```
Maybe instead of bt_codec_qos we could provide content type and target latency which later could be used by CAP to choose preferred QoS?

If we want to keep it this way I would call it target_qos, because we might end up with different QoS parameters based on the remote preference.

Thalley replied:
> Maybe instead of bt_codec_qos we could provide content type and target latency which later could be used by CAP to choose preferred QoS?

I originally designed it like that, but there are so many different values that go into the metadata and codec config that having an API that allows to set all these is near impossible, unless we split up the API to a `start_prepare` and a `start_commit` type of deal. It is much more flexible and simpler for the API to simply request the entire `bt_codec_qos` structure, and then verify the content (i.e. number of channels, etc.). 

> If we want to keep it this way I would call it target_qos, because we might end up with different QoS parameters based on the remote preference.

How? The remote preference is only a preference, and if the remote device actively rejects our QoS, then this procedure will fail. I don't think we should modify the requested QoS from the application autonomously.

rymanluk replied:
> How? The remote preference is only a preference, and if the remote device actively rejects our QoS, then this procedure will fail. I don't think we should modify the requested QoS from the application autonomously.

If we would not take audio server preference into account, it is not Nice but some Initiators can do it. But e.g. we need to take into account min/max presentation delay.

Thalley replied:
I'm not fully opposed to changing this to something that is a preference from the application, rather than a specific requirement. In that case, perhaps `bt_codec_qos_pref` could/should be used instead, but then we are exposing an API where the user cannot fully control the values, and I am not sure if that is a better or worse API that getting an error returned if the server rejects the request. 

@asbjornsabo Do you have any opinion?

Thalley commented on include/bluetooth/audio/cap.h:
```diff
@@ -86,6 +137,210 @@ struct bt_cap_cb {
+ *
+ * @return 0 if success, errno on failure.
+ */
+int bt_cap_unicast_audio_start(const struct bt_cap_unicast_audio_start_param *param,
```
Correct. Of course since there are multiple BAP procedures run on multiple devices, there are many places where this may fail.

Thalley commented on include/bluetooth/audio/cap.h:
```diff
@@ -86,6 +137,210 @@ struct bt_cap_cb {
+	/** The set is an ad-hoc set */
+	BT_CAP_SET_TYPE_AD_HOC,
+	/** The set is a CSIP Coordinated Set */
+	BT_CAP_SET_TYPE_CSIP,
```
Not sure what you mean here. 

CAP supports 2 sets: CSIP coordinated sets, and ad-hoc sets. 

What is the `GROUP` type?

rymanluk replied:
I think this is related to the below comment on how to represent group.

Thalley commented on include/bluetooth/audio/cap.h:
```diff
@@ -86,6 +137,210 @@ struct bt_cap_cb {
+/** Type of CAP set */
+enum bt_cap_set_type {
+	/** The set is an ad-hoc set */
+	BT_CAP_SET_TYPE_AD_HOC,
```
This just says that the sets supplied to the CAP type is an ad-hoc set, and not a coordinated set. Depending on the set type, CAP need to do the ordered access procedure from CSIP.

rymanluk replied:
Thanks. I think I figured that out later on and forget to remove this comment. `ad-hoc` means user came up with a group and this is not coordinated set.

Thalley commented on include/bluetooth/audio/cap.h:
```diff
@@ -86,6 +137,210 @@ struct bt_cap_cb {
+		 *
+		 * The number of pointers in this array shall be at least @p count
+		 */
+		struct bt_conn **members;
```
I'd rather want to avoid using `group` as a `group` in ISO (and also the Zephyr audio API) means something different from a set. 

If we want to expose an API to create and populate these groups/sets, then we should probably implement them as linked lists instead of arrays, but that can be done. 

Do you see any advantages of exposing an API to create these groups/sets instead of having it as a application allocated struct?

rymanluk replied:
linked list is fine. 

I think it would make things easier, if whenever user have to talk to stack in the context of group, it would be easier to have a defined group representation.
Otherwise, once user needs to construct array of members, in the `bt_audio_unicast_group` it has list of streams. That's only my feeling.

Thalley replied:
We could use list as a parameter instead of an array of pointers. This is a bit different from existing BT APIs that takes more than 1 pointer as input, but certainly not something I would be against. 

The downside of supporting the `slist` implementation, is that we will need to append `sys_snode_t node` to the structs that we want to put into a list. For something like `bt_conn` that is opaque, we will need to create a wrapper around it, but we could implement something like 

```c
struct bt_device_list_item {
    union {
        struct bt_conn *conn;
        struct bt_csis_client_set_member *member
    }

    sys_snode_t node;
}
```

So using a list approach will require a new struct, and then the application can use the `slist` API to create the list, such as 

```c

struct bt_device_list_item list_items[cnt] = { /* initialize array */ ); /* or just create these structs on the fly */

sys_slist_t list;

for (int i = 0; i < cnt; i++) {
    sys_slist_append(&list, &list_items[i].node);
}

bt_cap_unicast_audio_start_param(&list);
```

as opposed to the current solution:
```c
struct bt_conn conns[cnt] = { /* initialize array */ };

bt_cap_unicast_audio_start_param(&conns);
```

Thalley commented on include/bluetooth/audio/cap.h:
```diff
@@ -86,6 +137,210 @@ struct bt_cap_cb {
+			struct bt_csis_client_set_member **members;
+
+			/** Set information used to verify the set @p members */
+			struct bt_csis_client_set_info *set_info;
```
Correct. 

The difference is a `bt_csis_client_set_member` may have multiple sets, and just supplying a `bt_csis_client_set_member` doesn't indicate which sets CAP should read the rank (and possibly lock) value of, so there needs to be a specific set supplied as well.

rymanluk replied:
Ok. Do I understand that this `set_info` comes from one of the member?
Let's imagine each device have couple of CSIS instances, each in different context (included by different services).
How CAP will find the correct CAP instance? Do you expect it to be done based on SIRK and group size in the provided `set_info`?

Thalley replied:
> Ok. Do I understand that this `set_info` comes from one of the member? Let's imagine each device have couple of CSIS instances, each in different context (included by different services). How CAP will find the correct CAP instance? Do you expect it to be done based on SIRK and group size in the provided `set_info`?

There is only 1 CAS on each device, and CAP also restricts the number of CSIS to 1. I had briefly forgotten the last part, which would actually invalidate the need for the `set_info` here.

asbjornsabo commented on include/bluetooth/audio/cap.h:
```diff
@@ -0,0 +1,352 @@
+	 * @param conn        Connection object used in
+	 *                    bt_cap_find_ccid_service().
+	 * @param err         Error value. 0 on success, GATT error on positive
+	 *                    value or errno on negative value.
```
Inconsistent language.
Proposal: "Error value.  0 on success.  GATT error (positive values) or errno (negative value) on failure."
Or (shorter): "Error value.  0 on success.  GATT error or errno on failure."

(Also other places)

Thalley replied:
One thing we could do (and perhaps for the better) is to avoid ever returning GATT error codes in API functions, which would fix all the above issues. 

Will update.

Thalley replied:
Borrowed text from `bluetooth.h` and `gatt.h`. Please re-review

asbjornsabo commented on include/bluetooth/audio/cap.h:
```diff
@@ -0,0 +1,352 @@
+/**
+ * @brief Discovers audio support on a remote device.
+ *
+ * This will discover the Common Audio Service (CAS) on the remote device, to
```
Spelling ("to if")

asbjornsabo reviewed (commented):
(Some nitpicking while thinking more about the important stuff)

Thalley commented on subsys/bluetooth/audio/Kconfig.cap:
```diff
@@ -0,0 +1,51 @@
+	bool "Common Audio Service debug"
+	depends on BT_CAS
+	help
+	  Use this option to enable Common Audio Service debug logs for the
```
The reason I requested that, was that it was the OTS client which arguably isn't related to LE Audio, unlike this. 

I can remove it here, but then I suggest we do it for all the profiles that have it.

Thalley commented:
Closing in favor of the PRs that implements the API
